ast/src/parsed/display.rs

```diff
@@ -564,6 +564,53 @@ fn format_list<L: IntoIterator<Item = I>, I: Display>(list: L) -> String {
     format!("{}", list.into_iter().format(", "))
 }
 
+impl<E: Display> Expression<E> {
+    pub fn precedence(&self) -> Option<ExpressionPrecedence> {
+        match self {
+            Expression::BinaryOperation(_, op, _) => Some(op.precedence()),
+            _ => None,
+        }
+    }
+
+    pub fn format_binary_operation(&self, f: &mut Formatter<'_>) -> Result {
+        match self {
+            Expression::BinaryOperation(left, op, right) => {
+                let use_left_parenthes = match left.precedence() {
+                    Some(left_precedence) => {
+                        left_precedence > op.precedence()
+                            || (left_precedence == op.precedence()
+                                && op.associativity() == BinaryOperatorAssociativity::Right)
+                    }
+                    None => false,
```

Comment:
Shouldn't this be `true`? We don't know the precedence of the inner, so we should better put parentheses.

Maybe you are saying that it is not a binary operation, so we don't have to put parentheses, but then we should cross-check the Display function for Expression that it never puts parentheses for anything that returns None here.